### PR TITLE
Pessimistically locked SLOAD 

### PR DESCRIPTION
**Pull request by homakov**

### CODE REVIEW

**Comment by Arachnid** [comment]
Interesting idea!

I'm concerned, though, that as written this could make some useful patterns impossible - for instance, recursive calls to something that appends to an array, or transferring tokens to a contract that then immediately transfers (some of) them to another address.

A couple of alternate suggestions:
 - An explicit SLOCK operation
 - A change to gas accounting to only charge gas for the net change in storage at the end of a frame.

The latter would make mutexes and other concurrency primitives affordable, as well as other use cases that require temporary storage that persists across calls. Because none of this data gets written to disk, it's affordable for nodes, too.

**Comment by homakov** [comment]
What is needed to accept it for consideration?

Because i'd like to hear all opinions and try to figure out the best way out of it.

**Comment by GNSPS** [comment]
I was unsure of changing the default SLOAD behaviour but @Arachnid 's suggestion just lit up my eyes!

A dedicated operation for locking and only accounting for storage changes at the end of the frame is an awesome idea. Where do I sign? 😂

**Comment by homakov** [comment]
Well, effectively exposing properly SLOADUNSAFE + locked SLOAD would achieve the same result. I.e. if you need the reuse, just use the unsafe version (solidity may expose it as unsafestorage data type for example).

What i'm saying, it should be by default, and old code that people used to write should start working locked. All the use cases for recursive call of the same contract are corner cases and should be using another solidity functions. 

If anyone who's handy with eth blockchain, would be great if you could pass a global test whether there is anyone doing recursive calls except dao draining. Maybe I'm missing some case, don't know!

**Comment by GNSPS** [comment]
@homakov the 0x protocol core contracts are indeed making use of reentrancy, for example. Was that your question?

**Comment by MicahZoltu** [comment]
I'm very much against this EIP.  For a non-trivial dApp, touching one storage variable multiple times in a single transaction is not uncommon.  Also, since there are no races in the EVM, the situation is quite different from traditional DB locking as atomicity is already strongly guaranteed due to lack of parallelism.

I'm a fan of netting gas costs for storage updates, but that sounds like a separate EIP from this.  I wouldn't be against against an explicit lock opcode, thought I don't know if the value would outweigh the implementation cost.

**Comment by homakov** [comment]
@GNSPS can you point directly where it's used?

@MicahZoltu enlighten me, please, how non trivial dapps use same storage records recursively? I mean exact use case.

Another idea: SUNLOCK that unlocks particular SLOAD-ed key for future calls, and removes that key from $locked array. I.e. the opposite of SLOCK idea @Arachnid offered. 

Secure by default, easily unlockable if needed. Good?

**Comment by MicahZoltu** [comment]
Nick gave an example already that is pretty common.  Tokens are transferred into a contract early on in the transaction and then some portion (sometimes all) are transferred away later on in the transaction.  Other examples include mutexes that use a storage variable to prevent reentrancy like the [Open Zeppelin](https://github.com/OpenZeppelin/zeppelin-solidity/blob/master/contracts/ReentrancyGuard.sol) one.

Any changes to default behavior of SLOAD will break existing contracts, and therefore should not be included unless there is a _really_ compelling reason to do so.  I do not believe that this necessitates a breaking change to existing opcodes.

**Comment by homakov** [comment]
I still don't get it. Can you explain it with one by one Frame 1, frame 2, frame 3 scenario?

You probably assumed SLOAD is locked inside the same frame? That's pseudocode bug: in fact it is only locked for CALL frames inside current frame, and unlocked for everything else.

**Comment by homakov** [comment]
P.S. https://github.com/OpenZeppelin/zeppelin-solidity/blob/master/contracts/ReentrancyGuard.sol - not how security design works. Things like that must be on by default, otherwise they are never used. This EIP is a live example of a ton of such examples.

**Comment by MicahZoltu** [comment]
ERC223 makes it so token transfers call a function on the contract receiving the tokens.  This means that an ERC223 token transfer call stack may look like:
```
MyToken.transfer(FooContract, 1000); // msg.sender = BarContract
FooContract.onTokenReceived(MyToken, BarContract, 1000);
MyToken.transfer(AppleContract, 500);
```
While you may not _like_ the mechanism used in ReentrancyGuard, it exists in the wild today and will break if this change is implemented.

**Comment by homakov** [comment]
Now I see, the receiver sends parts of the tokens further to someone else, so they call same contract which has to SLOAD Foo's balance and would throw as it's locked by first frame.

SUNLOCK could be used here just after storage manipulations are done, explicitly. 
I.e. before onTokenReceived: balances.unlock(FooContract, BarContract)

That's a good example. But not sure why ReentrancyGuard would break? SLOAD will fail even sooner, so require() will not be executed, but the result will remain the same: fail on reentry.

If there's a erc223 style execution, new opcode behavior could break it. So it's better to be introduced for contracts created after the block... or the solution by Nick :)

**Comment by MicahZoltu** [comment]
Reentrancy guard would break because the lock is set to true before the method executes and then set to false after the method is finished executing.  This means the lock is set twice in the same transaction (first to true, then to false).

**Comment by homakov** [comment]
But there's nothing wrong with setting it twice. Global lock is only enforced in next frames, there's no point in disallowing multiple access to a storage record within same frame.

**Comment by pirapira** [comment]
Solidity has been using a single storage cell for multiple state variables (two `uint128`'s can fit in one word). So accessing a state variable might lock another.  The question is whether this troubles any existing Ethereum contracts.

**Comment by GNSPS** [comment]
Having a different opcode for it would mean that, in Solidity's specific case, newer contracts making use of this storage key locking feature could have the tight packing of state variables disabled.

**Comment by homakov** [comment]
@pirapira good catch. is it when solidity stores Structs only or literally two different vars could be stored under same record?

**Comment by pirapira** [comment]
@homakov the latter.  The more agressive one.

**Comment by Arachnid** [comment]
FWIW, my suggestion was that with net gas metering for storage changes, it wouldn't be necessary to introduce an explicit lock instruction; contracts could affordably implement locking themselves using storage variables - as well as a lot of other useful improvements.

I'm in agreement with the author that looking at transactions like database operations is a good model, thouh.

**Comment by GNSPS** [comment]
I do think that the explicit lock instruction has indeed got merit, though, since locking state would benefit from being implemented by default at the language level instead of being only a good practice. 🙂

**Comment by homakov** [comment]
Explicit not touching storage lock with next-to-0 gas (i guess you can't make it 0) - that's good enough for me.

**Comment by Arachnid** [comment]
@gnsps it's possible to offer this as a language primitive without a lock instruction.

**Comment by androlo** [comment]
Net gas calculation seems the simplest and most useful solution. With net gas cost, and modified pricing, it could make (affordable) temporary storage possible. Languages could implement it for example by adding a `lock` and `unlock` builtins for state variables (and auto generate temporary storage addresses for locks). They could even introduce a new type of variable (`transient` state variables perhaps) and auto generate code for clearing them at the end of any function where they are used. Also, as pointed out, it would only modify systems that are already in place. No new opcodes or anything like that. Fantastic.

File: EIPS/eip-lock_sload.md [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+The old_SLOAD might be added as a new opcode SLOADUNSAFE but can't see a single use case for it.
+
+## Test Cases
+
+The DAO + draining transaction = throw.
+
+## Implementations
+
+I can start working on ethereumjs-vm patch, if this EIP is approved.
+
```
Comment (pirapira) [inline EIPS/eip-lock_sload.md]: Please waive the copyright with CC0, as in https://github.com/ethereum/EIPs/blob/master/EIPS/eip-196.md#copyright

Comment (homakov) [reply EIPS/eip-lock_sload.md]: Sure - no copyright and CC0

File: EIPS/eip-lock_sload.md [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+EIP: X
+
+Title: Pessimistic locking for SLOAD
+
+Author: Egor Homakov <homakov@gmail.com>
+
+Status: Draft
+
+Type: Standards Track Core
```
Comment (pirapira) [inline EIPS/eip-lock_sload.md]:
```
Category: Core
```
should appear.

File: EIPS/eip-lock_sload.md [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+Re-entrancy is no different from a race condition: the same contract is invoked in the same frame stack, while they both can work on the same record from the table (not exactly in parallel, but frame 3 in the middle of frame 1). 
+
+And Ethereum's key-value contract storage, unlike traditional databases, _is all about finance_ and highly sensitive balances/amounts. On top of that the extra locking performance overhead doesn't cost anything for a blockchain and its use cases. 
+
+So since the storage is __a database that operates only with sensitive values__ - that's where we need isolation the most. Let's introduce __pessimistic locking by default__ and redefine the SLOAD opcode. That does not break a single valid usage of a contract, yet prevents DAO-like attacks including the DAO itself (__which will be considered a technical fork over a political one from now on__ if this EIP is accepted).
+
+## Specification
+
+Need your advices on this one. I'm sure the idea is outright obvious. The goal is to lock a key that was SLOAD-ed, i.e. turn `select value from Storage0xA where key="Key0xA"` into `select value from Storage0xA where key="Key0xA" SELECT FOR UPDATE`.
```
Comment (pirapira) [inline EIPS/eip-lock_sload.md]: Please make it explicit that `Storage0xA` stands for an account's storage.

File: EIPS/eip-lock_sload.md [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+Entire Storage0xA locking is probably an overkill. That would kill a few use-cases when the contract is called again in the same frame stack yet operates with different storage keys - no race condition bug there. Instead we should add a global lock array. After each frame termination the locks ackquired in this particular frame must be released.
```
Comment (pirapira) [inline EIPS/eip-lock_sload.md]: When there is an integrity condition among different storage keys, this approach is not guaranteed free of race condition bugs.

File: EIPS/eip-lock_sload.md [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+def SLOAD(account, key)
+  lock_id = account+key
+  if $locked.includes(lock_id)
+    throw;
+  else
+    $locked.push(account)
+    old_SLOAD(account, key)
+  end
+end
+
```
Comment (pirapira) [inline EIPS/eip-lock_sload.md]: What about `SSTORE`?  I guess `SSTORE` needs to be modified.  If that's not the case, that should be said explicitly I think.

File: EIPS/eip-lock_sload.md [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+It also certainly does not make it a vulnerability in Solidity. Otherwise, it would be fixed in Solidity already. It is not. What would the fix be? Locks. 
+
+Granular per-record locks inside Solidity would cost way too much gas - each SLOAD would have an overhead of another SLOAD/SSTORE on top of it. 
+
+Per-contract locks would also cost quite some gas and most contracts would opt-out from it. 
+
+## Backwards Compatibility
+
+The behavior should be changed from a certain block height. We should run a test how many transactions throughout entire history would throw with new consensus (wild guess: only the DAO one)
```
Comment (pirapira) [inline EIPS/eip-lock_sload.md]: My guess is like 20 more contracts would be affected.  I'm interested in the result.

File: EIPS/eip-lock_sload.md (hunk 5 above)
Comment (homakov) [reply EIPS/eip-lock_sload.md]: I'm busy with something else and in this threat there were better suggestions by @Arachnid  so this EIP isn't probably best way. I cannot keep it up to date - feel free to use create new one.

As of sstore, the safe way would be to throw error if the key is still locked, but most of the time you load before writing, so it would throw anyway.

Comment (pirapira) [reply EIPS/eip-lock_sload.md]: @homakov thank you for the response.  I marked this PR with `looking-for-champion`.

**Comment by Arachnid** [comment]
This is a courtesy notice to let you know that the format for EIPs has been modified slightly. If you want your draft merged, you will need to make some small changes to how your EIP is formatted:

 - Frontmatter is now contained between lines with only a triple dash ('---')
 - Headers in the frontmatter are now lowercase.

If your PR is editing an existing EIP rather than creating a new one, this has already been done for you, and you need only rebase your PR.

In addition, a continuous build has been setup, which will check your PR against the rules for EIP formatting automatically once you update your PR. This build ensures all required headers are present, as well as performing a number of other checks.

Please rebase your PR against the latest master, and edit your PR to use the above format for frontmatter. For convenience, here's a sample header you can copy and adapt:

```
---
eip: <num>
title: <title>
author: <author>
type: [Standards Track|Informational|Meta]
category: [Core|Networking|Interface|ERC] (for type: Standards Track only)
status: Draft
created: <date>
---
```

**Comment by jamesray1** [comment]
I'll share this with my Drops of Diamond team, working on sharding for Rust, as well as Parity. I'm too busy to work on this myself now, but I'll save it to Pocket.

**Comment by jamesray1** [comment]
> A change to gas accounting to only charge gas for the net change in storage at the end of a frame.
The latter would make mutexes and other concurrency primitives affordable, as well as other use cases that require temporary storage that persists across calls.

This would be particularly useful in the case of synchronous contract calls across shards. http://notes.ethereum.org/s/BJc_eGVFM#cross-shard-communication

**Comment by nicksavers** [comment]
@homakov The eip-X template has changed a bit. Please update your pull request to this for your draft to be merged. You can rename the file to `eip-718.md` and use this number in the preample.

**Comment by axic** [comment]
@homakov are you still pursuing this?

**Comment by homakov** [comment]
@axic I still think locking database would be very safe to do by default, but already 2 years passed without any changes in the protocol and/or solidity. As said in https://github.com/ethereum/EIPs/pull/718#discussion_r155156587 I don't pursue it anymore. 

It either needs a new champion (didn't happen in 2 years) or be closed.

File: EIPS/eip-lock_sload.md (hunk 2 above)
Comment (motafa1320) [reply EIPS/eip-lock_sload.md]: thanks
